File: Code/C++/longest_prefix_suffix.cpp
/*
-LONGEST PREFIX SUFFIX
-Given a string, find the length of longest substring that is both prefix and suffix
-overlapping of prefix and suffix not allowed

-idea is to divide the given string from between and check if both string are equal
-if so return the length of that substring
-otherwise check for the shorter length
 */

#include <iostream>
using namespace std;

int main()
{

    string s;
    // input from the user
    cin>>s;
    /*if length of given string is less
    than two, no such substring exist
    hence prints 0*/
    if (s.length()<2)
        cout<<0;
    else 
    {
        /*divide string into two equal parts
        and store starting indices of both*/
        int l=0, i=s.length()/2;
        /*loop calculating the length of
        such substring*/
        while (i<s.length()) 
        {
        /*check if left string character is equal
          to right one if so increament indices of both
          by 1*/
            if (s[i]==s[l])
            {
                ++l;
                ++i;
            }
        /*otherwise check for shoter substring*/
            else 
            {
                i=i-l+1;
                l=0;
            }
        }
        /*as overlapping is not allowed*/
        if (l>s.length()/2)
            cout<<s.length()/2;
        else
            cout<<l;
    }
    return 0;
}


/*

Test Cases

Test Case-1
 Input = "rrrrr"
 Output = 2

Test Case-2
 Input = "abcdabc"
 Output = 3

Time complexity = O(n), where n is length of string,i.e n=s.length()
Space complexity = O(1)

*/

File: Code/Python/longest_prefix_suffix.py
''' 

Prefix refers to a word at the beginning of another word, and suffix refers to a word added at the end of a word. Longest Prefix Suffix problem refers to finding the length of the longest prefix which is also suffix, of a string s, provided the prefix and suffix should not overlap. 

'''

#Longest prefix suffix function
def longest_prefix_suffix(s):
    length=len(s)
    #Iterate i through half the length of string s
    for i in range(length//2,0,-1): 
        #Prefix is from 0th index to ith index
        prefix=s[0:i]   
        #Suffix is the remaining part of string. That is, length minus prefix
        suffix=s[length-i:length] 
        #If prefix and suffix are equal, return length of string. Else, return 0
        if(prefix==suffix): 
            return i;
    return 0;

#Obtaining input from user
s=input()

#Calling the function
l=longest_prefix_suffix(s)

#Displaying output
print(l)

'''
 
Test Case :
 
Input : rrrrr
 
Output  : 2 

Input : hellohe
 
Output  : 2
 
Time Complexity: O(n)
 
Space Complexity : O(1)

'''
